packages/eslint-plugin/src/rules/use-unknown-in-catch-callback-variable.ts

```diff
@@ -289,44 +234,46 @@ export default createRule<[], MessageIds>({
     }
 
     return {
-      CallExpression(node): void {
-        if (node.arguments.length === 0 || !isPromiseCatchAccess(node.callee)) {
+      CallExpression({ arguments: args, callee }): void {
```

Comment:
I was playing around with this locally and I sketched out something like the following: 
<details>
<summary>Expand to see</summary> 

```ts
CallExpression(node): void {
  const { arguments: args, callee } = node;

  if (callee.type !== AST_NODE_TYPES.MemberExpression) {
    return;
  }

  const memberAccessKey = getStaticMemberAccessKey(callee);
  if (memberAccessKey == null) {
    return;
  }

  let promiseMethodInfo: {
    method: string;
    append: string;
    argToCheck: number;
  };

  if (memberAccessKey.value === 'catch') {
    promiseMethodInfo = { method: 'catch', append: '', argToCheck: 0 };
  } else if (memberAccessKey.value === 'then') {
    promiseMethodInfo = {
      method: 'then',
      append: ' rejection',
      argToCheck: 1,
    };
  } else {
    return;
  }

  // need to be enough args to check.
  if (args.length < promiseMethodInfo.argToCheck + 1) {
    return;
  }

  // Argument to check, and all arguments before it, are an "ordinary" argument (i.e. not a spread argument)
  // promise.catch(f), promise.catch(() => {}), promise.catch(<expression>, <<other-args>>)
  const argsToCheck = args.slice(0, promiseMethodInfo.argToCheck + 1);
  if (
    argsToCheck.some(arg => arg.type === AST_NODE_TYPES.SpreadElement)
  ) {
    return;
  }

  if (
    !tsutils.isThenableType(
      checker,
      services.esTreeNodeToTSNodeMap.get(callee),
      checker.getTypeAtLocation(
        services.esTreeNodeToTSNodeMap.get(callee.object),
      ),
    )
  ) {
    return;
  }

  const argToCheck = args[promiseMethodInfo.argToCheck] as Exclude<
    (typeof argsToCheck)[number],
    TSESTree.SpreadElement
  >;

  if (shouldFlagArgument(argToCheck)) {
    // We are now guaranteed to report, but we have a bit of work to do
    // to determine exactly where, and whether we can fix it.
    const overrides = refineReportForNormalArgumentIfPossible(argToCheck);
    context.report({
      node,
      messageId: 'useUnknown',
      data: {
        method: promiseMethodInfo.method,
        append: promiseMethodInfo.append,
      },
      ...overrides,
    });
  }
},

```
(with helper changed to)
```ts
/**
 * `x.memberName` => 'memberKey'
 *
 * `const mk = 'memberKey'; x[mk]` => 'memberKey'
 *
 * `const mk = 1234; x[mk]` => 1234
 */
function getStaticMemberAccessKey(
  memberExpression:
    | TSESTree.MemberExpressionComputedName
    | TSESTree.MemberExpressionNonComputedName,
  scope?: Scope.Scope | undefined,
): { value: unknown } | null {
  if (!memberExpression.computed) {
    // x.memberName case.
    return { value: memberExpression.property.name };
  }

  // x['memberName'] cases.
  return getStaticValue(memberExpression.property, scope);
}
```
</details>

By no means do I mean to say that this code is correct or fully polished, but I wanted to point out a few improvements that I think we should make:

1. Let's do as many syntactic checks as possible/reasonable before invoking the type checker (performance reasons). _The current changeset invokes the type checker on every method call before checking whether it's a `then` or `catch` access - we don't need to do that._
2. Let's really try to avoid doing overly complicated `.find()` stuff for what is really just two cases (`'then'` and `'catch'`). _Note - my code is sort of guilty of this too with the `.slice()`. Feel free to find a better way_.
3. No need to keep looking after having checked the rejection handler's position! 😁 
4. Eliminate the need to deal with the confusing `false | undefined | [number, { method: string, append: string }]` type on `promiserMethodInfo`.

Feel free to make this your own, and definitely I'm not implying you should just copy the code I wrote, but I'd like to clean up the impl a bit on these points.